### PR TITLE
react 18 props with children upgrade

### PR DESCRIPTION
**Pull request by RobotSail**
This PR aims to add PropsWithChildren explicitly on as many components as possible in order to make the React 18 upgrade easier.

Related to #18243 
#12252 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/core-components/src/components/Link/Link.test.tsx [hunk 1]
```diff
@@ -134,7 +134,7 @@ describe('<Link />', () => {
   });
 
   describe('useResolvedPath', () => {
-    const wrapper: WrapperComponent<{}> = ({ children }) => {
+    const wrapper: React.FC<React.PropsWithChildren<{}>> = ({ children }) => {
```
Comment (tudi2d) [inline packages/core-components/src/components/Link/Link.test.tsx:137]: Heads-up for the following [`ADR` on `React.FC`](https://backstage.io/docs/architecture-decisions/adrs-adr006)

**Review by tudi2d — CHANGES_REQUESTED** [review]
Thanks for the PR - definitly something to be fixed. I want to highlight the following alternative avoiding `PropsWithChildren` as we root for using `children?: React.ReactNode` in most places already (https://stackoverflow.com/a/71809927).

Other than that only a small change to remove `React.FC` required :)

File: packages/core-app-api/src/app/AppContext.test.tsx [hunk 2]
```diff
@@ -40,7 +40,7 @@ describe('v1 consumer', () => {
     };
 
     const renderedHook = renderHook(() => useMockAppV1(), {
-      wrapper: ({ children }) => (
+      wrapper: ({ children }: React.PropsWithChildren<{}>) => (
```
Comment (freben) [inline packages/core-app-api/src/app/AppContext.test.tsx:43]: It's unfortunate to put upon us as consumers to add this type, should have been added on the renderHook side instead :) Does switching over to the replacement library let us sidestep this? :) https://github.com/testing-library/react-hooks-testing-library#a-note-about-react-18-support

File: packages/test-utils/src/testUtils/appWrappers.tsx [hunk 3]
```diff
@@ -226,7 +231,7 @@ export function wrapInTestApp(
  * @public
  */
 export async function renderInTestApp(
-  Component: ComponentType | ReactNode,
+  Component: ComponentType<PropsWithChildren<{}>> | ReactNode,
```
Comment (freben) [inline packages/test-utils/src/testUtils/appWrappers.tsx:234]: This is actually not technically necessary, albeit not exactly harmful either. The test renderer just wants to render any component, it doesn't care about adding children to it.

**Review by freben — APPROVED** [review]
I'm good. @tudi2d how's your feeling on this one?

**Review by tudi2d — APPROVED** [review]
good work - let's ship it 🚢

**Pull request merged**